dask/order.py

```diff
@@ -115,787 +103,525 @@ def order(
+            # sometimes interested in the actual keys and the only way to
+            # benefit from the speedup of using integers would be to convert
+            # this back on demand which makes the code very hard to read.
+            roots.add(k)
+            result[k] = {k}
+            deps = dependents[k]
+            max_dependents[k] = len(deps)
+            for child in deps:
                 num_needed[child] -= 1
                 if not num_needed[child]:
-                    current_append(child)
-
+                    current.append(child)
     while current:
-        key = current_pop()
-        parents = dependents[key]
-        if len(parents) == 1:
-            (parent,) = parents
-            (
-                total_dependents,
-                min_dependencies,
-                max_dependencies,
-                min_heights,
-                max_heights,
-            ) = result[parent]
-            result[key] = (
-                1 + total_dependents,
-                min_dependencies,
-                max_dependencies,
-                1 + min_heights,
-                1 + max_heights,
-            )
+        key = current.pop()
+        for parent in dependents[key]:
+            num_needed[parent] -= 1
+            if not num_needed[parent]:
+                current.append(parent)
+        # At some point, all the roots are the same, particularly for dense
+        # graphs. We don't want to create new sets over and over again
+        new_set = None
+        identical_sets = True
+        if len(dependencies[key]) == 1:
+            child = next(iter(dependencies[key]))
+            max_dependents[key] = max_dependents[child]
+            result[key] = result[child]
         else:
-            (
-                total_dependents_,
-                min_dependencies_,
-                max_dependencies_,
-                min_heights_,
-                max_heights_,
-            ) = zip(*(result[parent] for parent in dependents[key]))
-            result[key] = (
-                1 + sum(total_dependents_),
-                min(min_dependencies_),
-                max(max_dependencies_),
-                1 + min(min_heights_),
-                1 + max(max_heights_),
-            )
-        for child in dependencies[key]:
-            num_needed[child] -= 1
-            if not num_needed[child]:
-                current_append(child)
-    return result
+            result_first = None
+
+            for child in dependencies[key]:
+                r_child = result[child]
+                if not result_first:
+                    result_first = r_child
+                    max_dependents[key] = max_dependents[child]
+                # This clause is written such that it can circuit break early
+                elif not (  # type: ignore[unreachable]
+                    identical_sets
+                    and (result_first is r_child or r_child.issubset(result_first))
+                ):
```

Comment:
@hendrikmakait I won't hold off merging for your review here but since we talked about it before this is the rewritten version. I think this much cleaner now (and I even wrote a couple of dedicated unit tests 🎉 )